sled-agent/src/storage/dump_setup.rs

```diff
@@ -0,0 +1,153 @@
+use crate::storage_manager::DiskWrapper;
+use camino::Utf8PathBuf;
+use illumos_utils::zpool::ZpoolHealth;
+use omicron_common::disk::DiskIdentity;
+use sled_hardware::DiskVariant;
+use slog::Logger;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::MutexGuard;
+
+#[derive(Default)]
+pub struct DumpSetup {
+    savecore_lock: Arc<std::sync::Mutex<()>>,
+}
+
+impl DumpSetup {
+    pub(crate) async fn poll_dumpdev_setup(
+        &self,
+        disks: &mut MutexGuard<'_, HashMap<DiskIdentity, DiskWrapper>>,
+        log: Logger,
+    ) {
+        let mut dump_slices = Vec::new();
+        let mut u2_dump_dirs = Vec::new();
+        for (_id, disk_wrapper) in disks.iter() {
+            match disk_wrapper {
+                DiskWrapper::Real { disk, .. } => match disk.variant() {
+                    DiskVariant::M2 => {
+                        match disk.dump_device_devfs_path(false) {
+                            Ok(path) => dump_slices.push(path),
+                            Err(err) => {
+                                warn!(log, "Error getting dump device devfs path: {err:?}");
+                            }
+                        }
+                    }
+                    DiskVariant::U2 => {
+                        let name = disk.zpool_name();
+                        if let Ok(info) = illumos_utils::zpool::Zpool::get_info(
+                            &name.to_string(),
+                        ) {
+                            if info.health() == ZpoolHealth::Online {
+                                u2_dump_dirs.push(name.dataset_mountpoint(
+                                    sled_hardware::disk::DUMP_DATASET,
+                                ));
+                            } else {
+                                warn!(log, "Zpool {name:?} not online, won't attempt to savecore dumps there");
+                            }
+                        }
+                    }
+                },
+                DiskWrapper::Synthetic { .. } => {}
+            }
+        }
+
+        dump_slices.sort();
+        u2_dump_dirs.sort();
+
+        let savecore_lock = self.savecore_lock.clone();
+        tokio::task::spawn_blocking(move || {
+            // TODO: a more reasonable way of deduplicating the effort.
+            let _guard = savecore_lock.lock();
```

Comment:
It looks like this lock exists to prevent us from running savecore concurrently, but we're configuring dumpadm in such as way that it should be running savecore on reboot, if there was a crash.

Do we risk racing between the "system-initiated savecore" and the "sled-agent-initiated savecore"?

Reply:
there's no system-initiated savecore at boot time, actually! i need to change the flags here to be more clear about that. at boot-time, our system has no configured dump slice until sled-agent sets one, and so won't ever invoke savecore by itself.